<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.4"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>singlepp: singlepp::Classifier Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">singlepp
   </div>
   <div id="projectbrief">A C++ library for cell type classification</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.4 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>singlepp</b></li><li class="navelem"><a class="el" href="classsinglepp_1_1Classifier.html">Classifier</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classsinglepp_1_1Classifier-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">singlepp::Classifier Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Automatically assign cell type labels based on an expression matrix.  
 <a href="classsinglepp_1_1Classifier.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="Classifier_8hpp_source.html">Classifier.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structsinglepp_1_1Classifier_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameters for annotation.  <a href="structsinglepp_1_1Classifier_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html">Prebuilt</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html" title="Prebuilt reference that can be directly used for annotation.">Prebuilt</a> reference that can be directly used for annotation.  <a href="structsinglepp_1_1Classifier_1_1Prebuilt.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structsinglepp_1_1Classifier_1_1PrebuiltIntersection.html">PrebuiltIntersection</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html" title="Prebuilt reference that can be directly used for annotation.">Prebuilt</a> reference requiring an intersection of features.  <a href="structsinglepp_1_1Classifier_1_1PrebuiltIntersection.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html">Results</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html" title="Results of the automated annotation.">Results</a> of the automated annotation.  <a href="structsinglepp_1_1Classifier_1_1Results.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a8da2466b32076a34b038d0bb88282ae1"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classsinglepp_1_1Classifier.html">Classifier</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a8da2466b32076a34b038d0bb88282ae1">set_quantile</a> (double q=<a class="el" href="structsinglepp_1_1Classifier_1_1Defaults.html#aa2b02289250da4ff08f1106a23850e14">Defaults::quantile</a>)</td></tr>
<tr class="separator:a8da2466b32076a34b038d0bb88282ae1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2a5cc599c01248cd5e14508a4b0ec0e5"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classsinglepp_1_1Classifier.html">Classifier</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a2a5cc599c01248cd5e14508a4b0ec0e5">set_fine_tune_threshold</a> (double t=<a class="el" href="structsinglepp_1_1Classifier_1_1Defaults.html#a6aa039661082457df6b283e38e696d67">Defaults::fine_tune_threshold</a>)</td></tr>
<tr class="separator:a2a5cc599c01248cd5e14508a4b0ec0e5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aed20bfa64b9dd557f38b665116e8005a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classsinglepp_1_1Classifier.html">Classifier</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#aed20bfa64b9dd557f38b665116e8005a">set_fine_tune</a> (bool f=<a class="el" href="structsinglepp_1_1Classifier_1_1Defaults.html#aa1690105051ae24ec10dc0b78c62bed3">Defaults::fine_tune</a>)</td></tr>
<tr class="separator:aed20bfa64b9dd557f38b665116e8005a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5406afbfe172ec57062f803d26c45f83"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classsinglepp_1_1Classifier.html">Classifier</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a5406afbfe172ec57062f803d26c45f83">set_top</a> (int t=<a class="el" href="structsinglepp_1_1Classifier_1_1Defaults.html#ac64a5e435923d56229cfc371861ffaa5">Defaults::top</a>)</td></tr>
<tr class="separator:a5406afbfe172ec57062f803d26c45f83"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7460114bf15720c699327f0f51d462da"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classsinglepp_1_1Classifier.html">Classifier</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a7460114bf15720c699327f0f51d462da">set_approximate</a> (bool a=<a class="el" href="structsinglepp_1_1Classifier_1_1Defaults.html#aa8668abed146d071d16b47f45d99413c">Defaults::approximate</a>)</td></tr>
<tr class="separator:a7460114bf15720c699327f0f51d462da"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a29c0f5ce6b3636f9dec21758aa4008b7"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html">Prebuilt</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a29c0f5ce6b3636f9dec21758aa4008b7">build</a> (const tatami::Matrix&lt; double, int &gt; *ref, const int *labels, <a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a> markers)</td></tr>
<tr class="separator:a29c0f5ce6b3636f9dec21758aa4008b7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9ebd4fe3a1653dbcbe9368ac90e0265f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a9ebd4fe3a1653dbcbe9368ac90e0265f">run</a> (const tatami::Matrix&lt; double, int &gt; *mat, const <a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html">Prebuilt</a> &amp;built, const int *mat_subset, int *best, std::vector&lt; double * &gt; &amp;scores, double *delta)</td></tr>
<tr class="separator:a9ebd4fe3a1653dbcbe9368ac90e0265f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afb890d0655b8de66dbdfe6c9cb784346"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#afb890d0655b8de66dbdfe6c9cb784346">run</a> (const tatami::Matrix&lt; double, int &gt; *mat, const <a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html">Prebuilt</a> &amp;built, int *best, std::vector&lt; double * &gt; &amp;scores, double *delta)</td></tr>
<tr class="separator:afb890d0655b8de66dbdfe6c9cb784346"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a44b45bcc689a72c21a0b4bb47044f8e5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a44b45bcc689a72c21a0b4bb47044f8e5">run</a> (const tatami::Matrix&lt; double, int &gt; *mat, const tatami::Matrix&lt; double, int &gt; *ref, const int *labels, <a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a> markers, int *best, std::vector&lt; double * &gt; &amp;scores, double *delta)</td></tr>
<tr class="separator:a44b45bcc689a72c21a0b4bb47044f8e5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab96aa9b5ecdde48afd71e754ba377ec5"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html">Results</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#ab96aa9b5ecdde48afd71e754ba377ec5">run</a> (const tatami::Matrix&lt; double, int &gt; *mat, const <a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html">Prebuilt</a> &amp;built)</td></tr>
<tr class="separator:ab96aa9b5ecdde48afd71e754ba377ec5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a801051fc515a2b90760392b256def1c0"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html">Results</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a801051fc515a2b90760392b256def1c0">run</a> (const tatami::Matrix&lt; double, int &gt; *mat, const <a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html">Prebuilt</a> &amp;built, const int *mat_subset)</td></tr>
<tr class="separator:a801051fc515a2b90760392b256def1c0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeee6da4d89dbbe90c51e9e98a7a3074c"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html">Results</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#aeee6da4d89dbbe90c51e9e98a7a3074c">run</a> (const tatami::Matrix&lt; double, int &gt; *mat, const tatami::Matrix&lt; double, int &gt; *ref, const int *labels, <a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a> markers)</td></tr>
<tr class="separator:aeee6da4d89dbbe90c51e9e98a7a3074c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acce2642df2b69740d675443a2c79ed6a"><td class="memTemplParams" colspan="2">template&lt;class Id &gt; </td></tr>
<tr class="memitem:acce2642df2b69740d675443a2c79ed6a"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structsinglepp_1_1Classifier_1_1PrebuiltIntersection.html">PrebuiltIntersection</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#acce2642df2b69740d675443a2c79ed6a">build</a> (size_t mat_nrow, const Id *mat_id, const tatami::Matrix&lt; double, int &gt; *ref, const Id *ref_id, const int *labels, <a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a> markers)</td></tr>
<tr class="separator:acce2642df2b69740d675443a2c79ed6a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5a16e1a5da6333f9a3cade28e20a6101"><td class="memTemplParams" colspan="2">template&lt;class Id &gt; </td></tr>
<tr class="memitem:a5a16e1a5da6333f9a3cade28e20a6101"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a5a16e1a5da6333f9a3cade28e20a6101">run</a> (const tatami::Matrix&lt; double, int &gt; *mat, const Id *mat_id, const tatami::Matrix&lt; double, int &gt; *ref, const Id *ref_id, const int *labels, <a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a> markers, int *best, std::vector&lt; double * &gt; &amp;scores, double *delta)</td></tr>
<tr class="separator:a5a16e1a5da6333f9a3cade28e20a6101"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a58a8370b958dcc93b81be7f7cc30b7cf"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a58a8370b958dcc93b81be7f7cc30b7cf">run</a> (const tatami::Matrix&lt; double, int &gt; *mat, const <a class="el" href="structsinglepp_1_1Classifier_1_1PrebuiltIntersection.html">PrebuiltIntersection</a> &amp;built, int *best, std::vector&lt; double * &gt; &amp;scores, double *delta)</td></tr>
<tr class="separator:a58a8370b958dcc93b81be7f7cc30b7cf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a74ad57bdf30b849951f99910810645c3"><td class="memTemplParams" colspan="2">template&lt;class Id &gt; </td></tr>
<tr class="memitem:a74ad57bdf30b849951f99910810645c3"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html">Results</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a74ad57bdf30b849951f99910810645c3">run</a> (const tatami::Matrix&lt; double, int &gt; *mat, const Id *mat_id, const tatami::Matrix&lt; double, int &gt; *ref, const Id *ref_id, const int *labels, <a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a> markers)</td></tr>
<tr class="separator:a74ad57bdf30b849951f99910810645c3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9ecf4c99790b850b43550420b6c7a2bd"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html">Results</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsinglepp_1_1Classifier.html#a9ecf4c99790b850b43550420b6c7a2bd">run</a> (const tatami::Matrix&lt; double, int &gt; *mat, const <a class="el" href="structsinglepp_1_1Classifier_1_1PrebuiltIntersection.html">PrebuiltIntersection</a> &amp;built)</td></tr>
<tr class="separator:a9ecf4c99790b850b43550420b6c7a2bd"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p >Automatically assign cell type labels based on an expression matrix. </p>
<p >This implements the <a href="https://bioconductor.org/packages/SingleR"><b>SingleR</b></a> algorithm for automated annotation of single-cell RNA-seq data. For each cell, we compute the Spearman rank correlation between that cell and the reference expression profiles. This is done using only the subset of genes that are label-specific markers, most typically the top genes from pairwise comparisons between each label's expression profiles. For each label, we take the correlations involving that label's reference profiles and convert it into a score. The label with the highest score is used as an initial label for that cell.</p>
<p >For each cell, we apply fine-tuning iterations to improve the label accuracy by refining the feature space. At each iteration, we find the subset of labels with scores that are close to the maximum score according to some threshold. We recompute the scores based on the markers for this label subset, and we repeat the process until only one label is left in the subset or the subset is unchanged. At the end of the iterations, the label with the highest score (or the only label, if just one is left) is used as the label for the cell. This process aims to remove noise by eliminating irrelevant genes when attempting to distinguish closely related labels.</p>
<p >Each label's score is defined as a user-specified quantile of the distribution of correlations across all reference profiles assigned to that label. (We typically consider a large quantile, e.g., the 80% percentile of the correlations.) The use of a quantile avoids problems with differences in the number of reference profiles per label; in contrast, just using the "top X correlations" would implicitly favor labels with more reference profiles.</p>
<p >The choice of Spearman's correlation provides some robustness against batch effects when comparing reference and test datasets. Only the relative expression <em>within</em> each cell needs to be comparable, not their relative expression across cells. As a result, it does not matter whether raw counts are supplied or log-transformed expression values, as the latter is a monotonic transformation of the latter (within each cell). The algorithm is also robust to differences in technologies between reference and test profiles, though it is preferable to have like-for-like comparisons.</p>
<dl class="section see"><dt>See also</dt><dd>Aran D et al. (2019). Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage. <em>Nat. Immunol.</em> 20, 163-172 </dd></dl>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a29c0f5ce6b3636f9dec21758aa4008b7" name="a29c0f5ce6b3636f9dec21758aa4008b7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a29c0f5ce6b3636f9dec21758aa4008b7">&#9670;&nbsp;</a></span>build() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html">Prebuilt</a> singlepp::Classifier::build </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>ref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int *&#160;</td>
          <td class="paramname"><em>labels</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a>&#160;</td>
          <td class="paramname"><em>markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">ref</td><td>Matrix for the reference expression profiles. Rows are genes while columns are samples. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">labels</td><td>An array of length equal to the number of columns of <code>ref</code>, containing the label for each sample. The smallest label should be 0 and the largest label should be equal to the total number of unique labels minus 1. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">markers</td><td>A vector of vectors of ranked marker genes for each pairwise comparison between labels, see <code>Markers</code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html" title="Prebuilt reference that can be directly used for annotation.">Prebuilt</a></code> instance that can be used in <code><a class="el" href="classsinglepp_1_1Classifier.html#a9ebd4fe3a1653dbcbe9368ac90e0265f">run()</a></code> for annotation of a test dataset. </dd></dl>

</div>
</div>
<a id="acce2642df2b69740d675443a2c79ed6a" name="acce2642df2b69740d675443a2c79ed6a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acce2642df2b69740d675443a2c79ed6a">&#9670;&nbsp;</a></span>build() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class Id &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structsinglepp_1_1Classifier_1_1PrebuiltIntersection.html">PrebuiltIntersection</a> singlepp::Classifier::build </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>mat_nrow</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Id *&#160;</td>
          <td class="paramname"><em>mat_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>ref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Id *&#160;</td>
          <td class="paramname"><em>ref_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int *&#160;</td>
          <td class="paramname"><em>labels</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a>&#160;</td>
          <td class="paramname"><em>markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Id</td><td>Gene identifier for each row.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat_nrow</td><td>Number of rows (genes) in the test dataset. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">mat_id</td><td>Pointer to an array of identifiers of length equal to <code>mat_nrow</code>. This should contain a unique identifier for each row of <code>mat</code> (typically a gene name or index). </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">ref</td><td>An expression matrix for the reference expression profiles, where rows are genes and columns are cells. This should have non-zero columns. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">ref_id</td><td>Pointer to an array of identifiers of length equal to the number of rows of any <code>ref</code>. This should contain a unique identifier for each row in <code>ref</code>, and should be comparable to <code>mat_id</code>. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">labels</td><td>An array of length equal to the number of columns of <code>ref</code>, containing the label for each sample. The smallest label should be 0 and the largest label should be equal to the total number of unique labels minus 1. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">markers</td><td>A vector of vectors of ranked marker genes for each pairwise comparison between labels, see <code>Markers</code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structsinglepp_1_1Classifier_1_1PrebuiltIntersection.html" title="Prebuilt reference requiring an intersection of features.">PrebuiltIntersection</a></code> instance that can be used in <code><a class="el" href="classsinglepp_1_1Classifier.html#a9ebd4fe3a1653dbcbe9368ac90e0265f">run()</a></code> for annotation of a test dataset with the same order of genes as specified in <code>mat_id</code>.</dd></dl>
<p>This method deals with the case where the genes are not in the same order and number across the test and reference datasets. It finds the intersection of genes and then prepares the references accordingly. </p>

</div>
</div>
<a id="a74ad57bdf30b849951f99910810645c3" name="a74ad57bdf30b849951f99910810645c3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a74ad57bdf30b849951f99910810645c3">&#9670;&nbsp;</a></span>run() <span class="overload">[1/10]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class Id &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html">Results</a> singlepp::Classifier::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Id *&#160;</td>
          <td class="paramname"><em>mat_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>ref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Id *&#160;</td>
          <td class="paramname"><em>ref_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int *&#160;</td>
          <td class="paramname"><em>labels</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a>&#160;</td>
          <td class="paramname"><em>markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Id</td><td>Gene identifier for each row.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Expression matrix of the test dataset, where rows are genes and columns are cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">mat_id</td><td>Pointer to an array of identifiers of length equal to the number of rows of <code>mat</code>. This should contain a unique identifier for each row of <code>mat</code> (typically a gene name or index). </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">ref</td><td>An expression matrix for the reference expression profiles, where rows are genes and columns are cells. This should have non-zero columns. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">ref_id</td><td>Pointer to an array of identifiers of length equal to the number of rows of any <code>ref</code>. This should contain a unique identifier for each row in <code>ref</code>, and should be comparable to <code>mat_id</code>. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">labels</td><td>An array of length equal to the number of columns of <code>ref</code>, containing the label for each sample. The smallest label should be 0 and the largest label should be equal to the total number of unique labels minus 1. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">markers</td><td>A vector of vectors of ranked marker genes for each pairwise comparison between labels, see <code>Markers</code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html" title="Results of the automated annotation.">Results</a></code> object containing the assigned labels and scores. </dd></dl>

</div>
</div>
<a id="a5a16e1a5da6333f9a3cade28e20a6101" name="a5a16e1a5da6333f9a3cade28e20a6101"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5a16e1a5da6333f9a3cade28e20a6101">&#9670;&nbsp;</a></span>run() <span class="overload">[2/10]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class Id &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void singlepp::Classifier::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Id *&#160;</td>
          <td class="paramname"><em>mat_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>ref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Id *&#160;</td>
          <td class="paramname"><em>ref_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int *&#160;</td>
          <td class="paramname"><em>labels</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a>&#160;</td>
          <td class="paramname"><em>markers</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>best</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; double * &gt; &amp;&#160;</td>
          <td class="paramname"><em>scores</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double *&#160;</td>
          <td class="paramname"><em>delta</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Id</td><td>Gene identifier for each row.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Expression matrix of the test dataset, where rows are genes and columns are cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">mat_id</td><td>Pointer to an array of identifiers of length equal to the number of rows of <code>mat</code>. This should contain a unique identifier for each row of <code>mat</code> (typically a gene name or index). </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">ref</td><td>An expression matrix for the reference expression profiles, where rows are genes and columns are cells. This should have non-zero columns. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">ref_id</td><td>Pointer to an array of identifiers of length equal to the number of rows of any <code>ref</code>. This should contain a unique identifier for each row in <code>ref</code>, and should be comparable to <code>mat_id</code>. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">labels</td><td>An array of length equal to the number of columns of <code>ref</code>, containing the label for each sample. The smallest label should be 0 and the largest label should be equal to the total number of unique labels minus 1. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">markers</td><td>A vector of vectors of ranked marker genes for each pairwise comparison between labels, see <code>Markers</code> for more details. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">best</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is filled with the index of the assigned label for each cell. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">scores</td><td>Vector of pointers of length equal to the number of labels. Each pointer should point to an array of length equal to the number of columns in <code>mat</code>. This is filled with the (non-fine-tuned) score for that label for each cell. Any pointer may be <code>NULL</code> in which case the scores for that label will not be saved. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">delta</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is filled with the difference between the highest and second-highest scores, possibly after fine-tuning. This may also be <code>NULL</code> in which case the deltas are not reported.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>best</code>, <code>scores</code> and <code>delta</code> are filled with their output values.</dd></dl>
<p>This version of <code><a class="el" href="classsinglepp_1_1Classifier.html#a9ebd4fe3a1653dbcbe9368ac90e0265f">run()</a></code> applies an intersection to find the common genes between <code>mat</code> and <code>ref</code>, based on their shared values in <code>mat_id</code> and <code>ref_id</code>. The annotation is then performed using only the subset of common genes. The aim is to easily accommodate differences in feature annotation between the test and reference profiles. </p>

</div>
</div>
<a id="ab96aa9b5ecdde48afd71e754ba377ec5" name="ab96aa9b5ecdde48afd71e754ba377ec5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab96aa9b5ecdde48afd71e754ba377ec5">&#9670;&nbsp;</a></span>run() <span class="overload">[3/10]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html">Results</a> singlepp::Classifier::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html">Prebuilt</a> &amp;&#160;</td>
          <td class="paramname"><em>built</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mat</td><td>Expression matrix of the test dataset, where rows are genes and columns are cells. If <code>already_subset = false</code>, this should have the same identity of genes as the reference matrices used in <code><a class="el" href="classsinglepp_1_1Classifier.html#a29c0f5ce6b3636f9dec21758aa4008b7">build()</a></code>; otherwise, each row should correspond to an element of <code>built.subset</code>. </td></tr>
    <tr><td class="paramname">built</td><td>An object produced by <code><a class="el" href="classsinglepp_1_1Classifier.html#a29c0f5ce6b3636f9dec21758aa4008b7">build()</a></code>.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html" title="Results of the automated annotation.">Results</a></code> object containing the assigned labels and scores. </dd></dl>

</div>
</div>
<a id="a801051fc515a2b90760392b256def1c0" name="a801051fc515a2b90760392b256def1c0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a801051fc515a2b90760392b256def1c0">&#9670;&nbsp;</a></span>run() <span class="overload">[4/10]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html">Results</a> singlepp::Classifier::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html">Prebuilt</a> &amp;&#160;</td>
          <td class="paramname"><em>built</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int *&#160;</td>
          <td class="paramname"><em>mat_subset</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Expression matrix of the test dataset, where rows are genes and columns are cells. This may have a different ordering of genes compared to the reference matrix used in <code><a class="el" href="classsinglepp_1_1Classifier.html#a29c0f5ce6b3636f9dec21758aa4008b7">build()</a></code>, provided that all genes corresponding to <code>built.subset</code> are present. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">built</td><td>An object produced by <code><a class="el" href="classsinglepp_1_1Classifier.html#a29c0f5ce6b3636f9dec21758aa4008b7">build()</a></code>. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">mat_subset</td><td>Pointer to an array of length equal to that of <code>built.subset</code>, containing the index of the row of <code>mat</code> corresponding to each gene in <code>built.subset</code>.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html" title="Results of the automated annotation.">Results</a></code> object containing the assigned labels and scores. </dd></dl>

</div>
</div>
<a id="a9ebd4fe3a1653dbcbe9368ac90e0265f" name="a9ebd4fe3a1653dbcbe9368ac90e0265f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9ebd4fe3a1653dbcbe9368ac90e0265f">&#9670;&nbsp;</a></span>run() <span class="overload">[5/10]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void singlepp::Classifier::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html">Prebuilt</a> &amp;&#160;</td>
          <td class="paramname"><em>built</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int *&#160;</td>
          <td class="paramname"><em>mat_subset</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>best</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; double * &gt; &amp;&#160;</td>
          <td class="paramname"><em>scores</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double *&#160;</td>
          <td class="paramname"><em>delta</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Expression matrix of the test dataset, where rows are genes and columns are cells. This may have a different ordering of genes compared to the reference matrix used to create <code>built</code>, provided that all genes corresponding to <code>built.subset</code> are present. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">built</td><td>An object produced by <code><a class="el" href="classsinglepp_1_1Classifier.html#a29c0f5ce6b3636f9dec21758aa4008b7">build()</a></code>. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">mat_subset</td><td>Pointer to an array of length equal to that of <code>built.subset</code>, containing the index of the row of <code>mat</code> corresponding to each gene in <code>built.subset</code>. That is, row <code>mat_subset[i]</code> in <code>mat</code> should be the same gene as row <code>built.subset[i]</code> in the reference matrix. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">best</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is filled with the index of the assigned label for each cell. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">scores</td><td>Vector of pointers to arrays of length equal to the number of columns in <code>mat</code>. This is filled with the (non-fine-tuned) score for each label for each cell. Any pointer may be <code>NULL</code> in which case the scores for that label will not be saved. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">delta</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is filled with the difference between the highest and second-highest scores, possibly after fine-tuning. This may also be <code>NULL</code> in which case the deltas are not reported. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">already_subset</td><td>Whether the rows of <code>mat</code> have already been subsetted to match <code>built.subset</code>.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>best</code>, <code>scores</code> and <code>delta</code> are filled with their output values. </dd></dl>

</div>
</div>
<a id="afb890d0655b8de66dbdfe6c9cb784346" name="afb890d0655b8de66dbdfe6c9cb784346"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afb890d0655b8de66dbdfe6c9cb784346">&#9670;&nbsp;</a></span>run() <span class="overload">[6/10]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void singlepp::Classifier::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structsinglepp_1_1Classifier_1_1Prebuilt.html">Prebuilt</a> &amp;&#160;</td>
          <td class="paramname"><em>built</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>best</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; double * &gt; &amp;&#160;</td>
          <td class="paramname"><em>scores</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double *&#160;</td>
          <td class="paramname"><em>delta</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Expression matrix of the test dataset, where rows are genes and columns are cells. This should have the same order and identity of genes as the reference matrix used to create <code>built</code>. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">built</td><td>An object produced by <code><a class="el" href="classsinglepp_1_1Classifier.html#a29c0f5ce6b3636f9dec21758aa4008b7">build()</a></code>. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">best</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is filled with the index of the assigned label for each cell. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">scores</td><td>Vector of pointers to arrays of length equal to the number of columns in <code>mat</code>. This is filled with the (non-fine-tuned) score for each label for each cell. Any pointer may be <code>NULL</code> in which case the scores for that label will not be saved. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">delta</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is filled with the difference between the highest and second-highest scores, possibly after fine-tuning. This may also be <code>NULL</code> in which case the deltas are not reported. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">already_subset</td><td>Whether the rows of <code>mat</code> have already been subsetted to match <code>built.subset</code>.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>best</code>, <code>scores</code> and <code>delta</code> are filled with their output values. </dd></dl>

</div>
</div>
<a id="a9ecf4c99790b850b43550420b6c7a2bd" name="a9ecf4c99790b850b43550420b6c7a2bd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9ecf4c99790b850b43550420b6c7a2bd">&#9670;&nbsp;</a></span>run() <span class="overload">[7/10]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html">Results</a> singlepp::Classifier::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structsinglepp_1_1Classifier_1_1PrebuiltIntersection.html">PrebuiltIntersection</a> &amp;&#160;</td>
          <td class="paramname"><em>built</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mat</td><td>Expression matrix of the test dataset, where rows are genes and columns are cells. </td></tr>
    <tr><td class="paramname">built</td><td>An object produced by <code><a class="el" href="classsinglepp_1_1Classifier.html#a29c0f5ce6b3636f9dec21758aa4008b7">build()</a></code> with intersections.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html" title="Results of the automated annotation.">Results</a></code> object containing the assigned labels and scores. </dd></dl>

</div>
</div>
<a id="a58a8370b958dcc93b81be7f7cc30b7cf" name="a58a8370b958dcc93b81be7f7cc30b7cf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a58a8370b958dcc93b81be7f7cc30b7cf">&#9670;&nbsp;</a></span>run() <span class="overload">[8/10]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void singlepp::Classifier::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structsinglepp_1_1Classifier_1_1PrebuiltIntersection.html">PrebuiltIntersection</a> &amp;&#160;</td>
          <td class="paramname"><em>built</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>best</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; double * &gt; &amp;&#160;</td>
          <td class="paramname"><em>scores</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double *&#160;</td>
          <td class="paramname"><em>delta</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Expression matrix of the test dataset, where rows are genes and columns are cells. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">built</td><td>An object produced by <code><a class="el" href="classsinglepp_1_1Classifier.html#a29c0f5ce6b3636f9dec21758aa4008b7">build()</a></code> with intersections. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">best</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is filled with the index of the assigned label for each cell. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">scores</td><td>Vector of pointers to arrays of length equal to the number of columns in <code>mat</code>. This is filled with the (non-fine-tuned) score for each label for each cell. Any pointer may be <code>NULL</code> in which case the scores for that label will not be saved. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">delta</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is filled with the difference between the highest and second-highest scores, possibly after fine-tuning. This may also be <code>NULL</code> in which case the deltas are not reported.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>best</code>, <code>scores</code> and <code>delta</code> are filled with their output values. </dd></dl>

</div>
</div>
<a id="aeee6da4d89dbbe90c51e9e98a7a3074c" name="aeee6da4d89dbbe90c51e9e98a7a3074c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aeee6da4d89dbbe90c51e9e98a7a3074c">&#9670;&nbsp;</a></span>run() <span class="overload">[9/10]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html">Results</a> singlepp::Classifier::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>ref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int *&#160;</td>
          <td class="paramname"><em>labels</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a>&#160;</td>
          <td class="paramname"><em>markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Expression matrix of the test dataset, where rows are genes and columns are cells. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">ref</td><td>An expression matrix for the reference expression profiles. This should have non-zero columns and the same number of rows (i.e., genes) at <code>mat</code>. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">labels</td><td>An array of length equal to the number of columns of <code>ref</code>, containing the label for each sample. The smallest label should be 0 and the largest label should be equal to the total number of unique labels minus 1. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">markers</td><td>A vector of vectors of ranked marker genes for each pairwise comparison between labels, see <code>Markers</code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structsinglepp_1_1Classifier_1_1Results.html" title="Results of the automated annotation.">Results</a></code> object containing the assigned labels and scores. </dd></dl>

</div>
</div>
<a id="a44b45bcc689a72c21a0b4bb47044f8e5" name="a44b45bcc689a72c21a0b4bb47044f8e5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a44b45bcc689a72c21a0b4bb47044f8e5">&#9670;&nbsp;</a></span>run() <span class="overload">[10/10]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void singlepp::Classifier::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const tatami::Matrix&lt; double, int &gt; *&#160;</td>
          <td class="paramname"><em>ref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int *&#160;</td>
          <td class="paramname"><em>labels</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="Markers_8hpp.html#ae39cf61993000d182cb3464b4ace1406">Markers</a>&#160;</td>
          <td class="paramname"><em>markers</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>best</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; double * &gt; &amp;&#160;</td>
          <td class="paramname"><em>scores</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double *&#160;</td>
          <td class="paramname"><em>delta</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Expression matrix of the test dataset, where rows are genes and columns are cells. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">ref</td><td>An expression matrix for the reference expression profiles. This should have non-zero columns and the same number of rows (i.e., genes) at <code>mat</code>. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">labels</td><td>An array of length equal to the number of columns of <code>ref</code>, containing the label for each sample. The smallest label should be 0 and the largest label should be equal to the total number of unique labels minus 1. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">markers</td><td>A vector of vectors of ranked marker genes for each pairwise comparison between labels, see <code>Markers</code> for more details. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">best</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is filled with the index of the assigned label for each cell. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">scores</td><td>Vector of pointers of length equal to the number of labels. Each pointer should point to an array of length equal to the number of columns in <code>mat</code>. This is filled with the (non-fine-tuned) score for that label for each cell. Any pointer may be <code>NULL</code> in which case the scores for that label will not be saved. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">delta</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is filled with the difference between the highest and second-highest scores, possibly after fine-tuning. This may also be <code>NULL</code> in which case the deltas are not reported.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>best</code>, <code>scores</code> and <code>delta</code> are filled with their output values. </dd></dl>

</div>
</div>
<a id="a7460114bf15720c699327f0f51d462da" name="a7460114bf15720c699327f0f51d462da"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7460114bf15720c699327f0f51d462da">&#9670;&nbsp;</a></span>set_approximate()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classsinglepp_1_1Classifier.html">Classifier</a> &amp; singlepp::Classifier::set_approximate </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>a</em> = <code><a class="el" href="structsinglepp_1_1Classifier_1_1Defaults.html#aa8668abed146d071d16b47f45d99413c">Defaults::approximate</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">a</td><td>Whether to use an approximate method to quickly find the quantile. This sacrifices some accuracy for speed when labels have many reference profiles.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classsinglepp_1_1Classifier.html" title="Automatically assign cell type labels based on an expression matrix.">Classifier</a></code> object. </dd></dl>

</div>
</div>
<a id="aed20bfa64b9dd557f38b665116e8005a" name="aed20bfa64b9dd557f38b665116e8005a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aed20bfa64b9dd557f38b665116e8005a">&#9670;&nbsp;</a></span>set_fine_tune()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classsinglepp_1_1Classifier.html">Classifier</a> &amp; singlepp::Classifier::set_fine_tune </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>f</em> = <code><a class="el" href="structsinglepp_1_1Classifier_1_1Defaults.html#aa1690105051ae24ec10dc0b78c62bed3">Defaults::fine_tune</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>Whether to perform fine-tuning. This can be disabled to improve speed at the cost of accuracy.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classsinglepp_1_1Classifier.html" title="Automatically assign cell type labels based on an expression matrix.">Classifier</a></code> object. </dd></dl>

</div>
</div>
<a id="a2a5cc599c01248cd5e14508a4b0ec0e5" name="a2a5cc599c01248cd5e14508a4b0ec0e5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2a5cc599c01248cd5e14508a4b0ec0e5">&#9670;&nbsp;</a></span>set_fine_tune_threshold()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classsinglepp_1_1Classifier.html">Classifier</a> &amp; singlepp::Classifier::set_fine_tune_threshold </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>t</em> = <code><a class="el" href="structsinglepp_1_1Classifier_1_1Defaults.html#a6aa039661082457df6b283e38e696d67">Defaults::fine_tune_threshold</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">t</td><td>Threshold to use to select the top-scoring subset of labels during fine-tuning. Larger values increase the chance of recovering the correct label at the cost of computational time.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classsinglepp_1_1Classifier.html" title="Automatically assign cell type labels based on an expression matrix.">Classifier</a></code> object.</dd></dl>
<p>Needless to say, one should not set <code>t</code> to a value that is too large. Otherwise, the first fine-tuning iteration would just contain all labels and there would be no reduction of the marker space. </p>

</div>
</div>
<a id="a8da2466b32076a34b038d0bb88282ae1" name="a8da2466b32076a34b038d0bb88282ae1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8da2466b32076a34b038d0bb88282ae1">&#9670;&nbsp;</a></span>set_quantile()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classsinglepp_1_1Classifier.html">Classifier</a> &amp; singlepp::Classifier::set_quantile </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>q</em> = <code><a class="el" href="structsinglepp_1_1Classifier_1_1Defaults.html#aa2b02289250da4ff08f1106a23850e14">Defaults::quantile</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">q</td><td>Quantile to use to compute a per-label score from the correlations.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classsinglepp_1_1Classifier.html" title="Automatically assign cell type labels based on an expression matrix.">Classifier</a></code> object.</dd></dl>
<p>Values of <code>q</code> closer to 0.5 focus on the behavior of the majority of a label's reference profiles. Smaller values will be more sensitive to the presence of a subset of profiles that are more similar to the test cell, which can be useful when the reference profiles themselves are heterogeneous. </p>

</div>
</div>
<a id="a5406afbfe172ec57062f803d26c45f83" name="a5406afbfe172ec57062f803d26c45f83"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5406afbfe172ec57062f803d26c45f83">&#9670;&nbsp;</a></span>set_top()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classsinglepp_1_1Classifier.html">Classifier</a> &amp; singlepp::Classifier::set_top </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>t</em> = <code><a class="el" href="structsinglepp_1_1Classifier_1_1Defaults.html#ac64a5e435923d56229cfc371861ffaa5">Defaults::top</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">t</td><td>Number of top markers to use from each pairwise comparison between labels. Larger values improve the stability of the correlations at the cost of increasing noise and computational work.</td></tr>
  </table>
  </dd>
</dl>
<p>Setting it to a negative value will instruct <code><a class="el" href="classsinglepp_1_1Classifier.html#a9ebd4fe3a1653dbcbe9368ac90e0265f">run()</a></code> to use all supplied markers. This is useful in situations where the supplied markers have already been curated.</p>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classsinglepp_1_1Classifier.html" title="Automatically assign cell type labels based on an expression matrix.">Classifier</a></code> object. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>singlepp/<a class="el" href="Classifier_8hpp_source.html">Classifier.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.4
</small></address>
</body>
</html>
